SimCalorimetry/EcalSimProducers/src/EcalDigiProducer_Ph2.cc

```diff
@@ -0,0 +1,336 @@
+
+  randomEngine_ = nullptr;  // to prevent access outside event
+}
+
+void EcalDigiProducer_Ph2::beginLuminosityBlock(edm::LuminosityBlock const& lumi, edm::EventSetup const& setup) {
+  edm::Service<edm::RandomNumberGenerator> rng;
+  if (!rng.isAvailable()) {
+    throw cms::Exception("Configuration") << "RandomNumberGenerator service is not available.\n"
+                                             "You must add the service in the configuration file\n"
+                                             "or remove the module that requires it.";
+  }
+  CLHEP::HepRandomEngine* engine = &rng->getEngine(lumi.index());
+
+  if (nullptr != m_APDResponse)
+    m_APDResponse->initialize(engine);
+  m_EBResponse->initialize(engine);
+}
+
+void EcalDigiProducer_Ph2::checkCalibrations(const edm::Event& event, const edm::EventSetup& eventSetup) {
+  // Pedestals from event setup
+
+  edm::ESHandle<EcalLiteDTUPedestalsMap> dbPed;
+  eventSetup.get<EcalLiteDTUPedestalsRcd>().get(dbPed);
+  const EcalLiteDTUPedestalsMap* pedestals(dbPed.product());
+
+  m_Coder->setPedestals(pedestals);
+  if (nullptr != m_APDCoder)
+    m_APDCoder->setPedestals(pedestals);
+
+  // Ecal Intercalibration Constants
+  edm::ESHandle<EcalIntercalibConstantsMC> pIcal;
+  eventSetup.get<EcalIntercalibConstantsMCRcd>().get(pIcal);
+  const EcalIntercalibConstantsMC* ical(pIcal.product());
+
+  m_Coder->setIntercalibConstants(ical);
+  if (nullptr != m_APDCoder)
+    m_APDCoder->setIntercalibConstants(ical);
+
+  m_EBResponse->setIntercal(ical);
+  if (nullptr != m_APDResponse)
+    m_APDResponse->setIntercal(ical);
+
+  // Ecal LaserCorrection Constants
+  edm::ESHandle<EcalLaserDbService> laser;
+  eventSetup.get<EcalLaserDbRecord>().get(laser);
+  const edm::TimeValue_t eventTimeValue = event.time().value();
+
+  m_EBResponse->setEventTime(eventTimeValue);
+  m_EBResponse->setLaserConstants(laser.product(), m_useLCcorrection);
+
+  // ADC -> GeV Scale
+  edm::ESHandle<EcalADCToGeVConstant> pAgc;
+  eventSetup.get<EcalADCToGeVConstantRcd>().get(pAgc);
+  const EcalADCToGeVConstant* agc = pAgc.product();
+
+  m_Coder->setGainRatios(ecalPh2::gains[0] / ecalPh2::gains[1]);
+  if (nullptr != m_APDCoder)
+    m_APDCoder->setGainRatios(ecalPh2::gains[0] / ecalPh2::gains[1]);
+
+  const double EBscale((agc->getEBValue()) * ecalPh2::gains[1] * (ecalPh2::MAXADC)*m_EBs25notCont);
+
+  LogDebug("EcalDigi") << " GeV/ADC = " << agc->getEBValue() << "\n"
+                       << " notCont = " << m_EBs25notCont << "\n"
+                       << " saturation for EB = " << EBscale << ", " << m_EBs25notCont;
+
+  m_Coder->setFullScaleEnergy(EBscale);
+  if (nullptr != m_APDCoder)
+    m_APDCoder->setFullScaleEnergy(EBscale);
+}
+
+void EcalDigiProducer_Ph2::checkGeometry(const edm::EventSetup& eventSetup) {
+  // TODO find a way to avoid doing this every event
+  edm::ESHandle<CaloGeometry> hGeometry;
+  eventSetup.get<CaloGeometryRecord>().get(hGeometry);
```

Comment:
See https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit

Reply:
Is there some difficulty fixing this part?